sourceinfo/source_code_info.go

```diff
@@ -580,43 +613,95 @@ func (sci *sourceCodeInfo) getTrailingComments(n ast.Node) comments {
+		if trail.Len() == 0 {
+			trail, detached = sci.maybeDonate(prevInfo, info, detached)
+		}
+	} else {
+		trail = emptyComments
 	}
-	if prevInfo.End().Line == info.Start().Line {
-		// if the tokens are on the same line, no way to attribute
-		return trail, nil
+	detached, lead := sci.maybeAttach(prevInfo, info, trail.Len() > 0, detached)
+	return trail, detached, lead
+}
+
+func (sci *sourceCodeInfo) maybeDonate(prevInfo ast.NodeInfo, info ast.NodeInfo, lead []comments) (t comments, l []comments) {
+	if len(lead) == 0 {
+		// nothing to donate
+		return emptyComments, nil
+	}
+	if !sci.extraComments {
+		// Mirroring protoc for now: if tokens on the same line, attribution
+		// is ambiguous so drop everything.
+		// This condition can be dropped (and we can stop dropping comments)
+		// once this PR is released:
+		//   https://github.com/protocolbuffers/protobuf/pull/10660
+		if prevInfo.End().Line == info.Start().Line {
+			return emptyComments, nil
+		}
 	}
 	if lead[0].Index(0).Start().Line > prevInfo.End().Line+1 {
 		// first comment is detached from previous token, so can't be a trailing comment
-		return trail, lead
+		return emptyComments, lead
 	}
 	if len(lead) > 1 {
 		// multiple groups? then donate first comment to previous token
 		return lead[0], lead[1:]
 	}
-
-	if lead[0].Index(lead[0].Len()-1).End().Line < info.Start().Line-1 {
+	// there is only one element in lead
+	comment := lead[0]
+	if comment.Index(comment.Len()-1).End().Line < info.Start().Line-1 {
 		// there is a blank line between the comments and subsequent token, so
 		// we can donate the comment to previous token
-		return lead[0], lead[1:]
+		return comment, nil
 	}
-	if txt := info.RawText(); len(txt) == 1 && strings.Contains("}])", txt) {
+	if txt := info.RawText(); len(txt) == 1 && strings.Contains("}]),;", txt) {
 		// token is a symbol for the end of a scope, which doesn't need a leading comment
-		return lead[0], lead[1:]
+		return comment, nil
+	}
+
+	// cannot donate
+	return emptyComments, lead
+}
+
+func (sci *sourceCodeInfo) maybeAttach(prevInfo ast.NodeInfo, info ast.NodeInfo, hasTrail bool, lead []comments) (d []comments, l comments) {
+	if len(lead) == 0 {
+		return nil, emptyComments
+	}
+
+	if len(lead) == 1 && !hasTrail && prevInfo.IsValid() {
+		// If the one comment appears attached to both previous and next tokens,
+		// don't attach to either.
+		comment := lead[0]
+		attachedToPrevious := comment.Index(0).Start().Line == prevInfo.End().Line
+		attachedToNext := comment.Index(comment.Len()-1).End().Line == info.Start().Line
+		if attachedToPrevious && attachedToNext {
+			if !sci.extraComments {
+				// Mirroring protoc for now: if comment starts on line of previous token
+				// and ends on line of next, attribution is ambiguous so drop everything.
+				// This condition can be dropped (and we can stop dropping comments)
+				// once this PR is released:
+				//   https://github.com/protocolbuffers/protobuf/pull/10660
```

Comment:
I was surprised at how quickly this PR was merged. Hopefully it doesn't end up getting reverted like my last two PRs :/

This will involve not only changes to this area of `protocompile` but also to the language spec, since the behavior is changing: previously, if it was unclear/ambiguous to what token to attach a comment, the comment would be dropped. Now the comments are not dropped but retained as detached (neither trailing nor leading), exactly as we're doing for the "extra comments" mode here.